### PR TITLE
Issue/51 gsp power

### LINKED ISSUE #51: Predict GSP-level PV power
Background: "GSP" means "grid supply point".  National Grid divide Britain up into ~350 non-overlapping regions, and National Grid need PV forecasts for each of these regions.  Also, National Grid will probably expect our GSP-level PV forecasts to be calibrated to Sheffield Solar's PV Live "ground truth" for PV generation at each GSP.

Several options:

1) ML predicts PV for each of the ~ 1 million PV systems in Britain.  Then we'll sum spatially over each GSP region, and then calibrate to Sheffield Solar's PV Live, perhaps using a Temporal Fusion Transformer.  We'll need the location & capacity for each of those million PV systems.  Not entirely clear how to produce probabilistic GSP-level forecasts except Monte Carlo sampling from the PV-level probability distributions?  Maybe we could deliver two forecasts: one that's calibrated to PV Live, and one that isn't?

2) ML directly predicts PV power for each GSP.  Almost certainly still want to train on individual PV systems (as well as GSPs).  Maybe it's as simple as including the GSPs a separate "PV system IDs" and feeding them into the "PV system ID" embedding.  Or maybe have two embeddings (to make it really clear to the net that individual PV systems and GSPs are two separate concepts.  Although a small GSP might not be much larger than a large, multi-field PV farm, so the concepts aren't _that_ different).  This is nice because it directly outputs exactly what we need (GSP-level PV, nicely calibrated to PV Live) and makes it trivial to produce probabilistic forecasts.  But do the large GSPs fit within the ML's square of satellite imagery?   (If not, maybe drop the spatial res and enlarge the spatial extent?)

3) predict PV yield over a regular (4km?) grid spacing over Britain. Then split those predictions per GSP. Feed into a simple fully connected net (one per GSP) which predicts GSP-level PV according to PV Live. 

4) OCF predicts PV power for the ~20,000 PassivSystem PV systems that Sheffield Solar use as input to PV Live.  Sheffield Solar use their PV Live algorithm to upscale from those 20,000 PV systems to the total output per GSP.  As an added bonus, OCF could try to predict *5-minutely* data for the ~20,000 PV systems which, actually, only report half-hourly data once-per-day.

Could provide a 'mask' to the net, which shows the spatial extent of the GSP (and shows a point for small PV systems; and geometry for larger PV farms???  Although few if any PV farms are larger than a pixel of satellite imagery)

To do this stuff, we need some more data:

- [ ] Sheffield Solar's historical GSP-level PV Live: https://github.com/openclimatefix/nowcasting_dataset/issues/88
- [ ] Recent satellite imagery and PV data: https://github.com/openclimatefix/nowcasting_dataset/issues/95

### PR DESCRIPTION
**Pull request by peterdudfield**
# Pull Request

## Description

- Adjust code to predict gsp_power (and option to predict pv output)
- Add convolution layers on NWP data 

Does not cover
- #85 
- #84 

Fixes issue #51 

## How Has This Been Tested?

- Unittest have been adapted and added
- experiments
- 1. https://app.neptune.ai/o/OpenClimateFix/org/predict-pv-yield/e/PRED-283/charts
- 2. https://app.neptune.ai/o/OpenClimateFix/org/predict-pv-yield/e/PRED-314/charts
- 3. https://app.neptune.ai/o/OpenClimateFix/org/predict-pv-yield/e/PRED-320/charts

- [ ] No
- [x] Yes

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/nowcasting/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: experiments/2021-09/2021-09-24/experiments.txt [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# Baseline
+
+Ran baseline line on new v6 GCP dataset
```
Comment (JackKelly) [inline experiments/2021-09/2021-09-24/experiments.txt]:
```suggestion
Ran baseline on new v6 GCP dataset
```

Also, please could you briefly describe what the baseline model is?

File: experiments/2021-09/2021-09-24/experiments.txt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+https://app.neptune.ai/o/OpenClimateFix/org/predict-pv-yield/e/PRED-283/monitoring
+
+Takes about 4 minutes to run the validation epoch
+
+This is just for forecast 1 timestep in the future
+MAE = 0.0562
+
+# Conv3d
+
+ran not using nwp data, or datetimte features
```
Comment (JackKelly) [inline experiments/2021-09/2021-09-24/experiments.txt]:
```suggestion
ran not using nwp data, or datetime features
```

File: experiments/2021-09/2021-09-27/experiments.txt [hunk 3]
```diff
@@ -0,0 +1,25 @@
+# Baseline
+
+Ran baseline line on new v6 GCP dataset
```
Comment (JackKelly) [inline experiments/2021-09/2021-09-27/experiments.txt]:
```suggestion
Ran baseline on new v6 GCP dataset
```

Or have I misunderstood?!

File: experiments/2021-09/2021-09-24/experiments.txt [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+https://app.neptune.ai/o/OpenClimateFix/org/predict-pv-yield/e/PRED-314/monitoring
+
+MAE = 0.0401
```
Comment (JackKelly) [inline experiments/2021-09/2021-09-24/experiments.txt:17]: Nice!

File: experiments/2021-09/2021-09-27/experiments.txt [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+# Conv3d
+
+ran not using nwp data, or datetimte features
+https://app.neptune.ai/o/OpenClimateFix/org/predict-pv-yield/e/PRED-314/monitoring
+
+MAE = 0.0401
+
+# Conv3d (Sat and NWP)
+
+Using both sat and nwp into tow seperate convoultion nets.
```
Comment (JackKelly) [inline experiments/2021-09/2021-09-27/experiments.txt]:
```suggestion
Using both sat and nwp into two separate convolution nets.
```

File: predict_pv_yield/data/dataloader.py [hunk 6]
```diff
@@ -123,13 +123,15 @@ def test_dataloader(self):
 class FakeDataset(torch.utils.data.Dataset):
     """Fake dataset."""
 
-    def __init__(self, batch_size=32, seq_length=19, width=16, height=16, number_sat_channels=8, length=10):
+    def __init__(self, batch_size=32, seq_length_5=19, seq_length_30=4, width=16, height=16, number_sat_channels=8, length=10, nwp_image_size=2):
```
Comment (JackKelly) [inline predict_pv_yield/data/dataloader.py:126]: Instead of passing in `seq_length_5` and `seq_length_30`, should we instead pass in `history_minutes` and `forecast_minutes` and let the code figure out `seq_length_5` and `seq_length_30`?  Or are there times when the user will want fine-grain control over `seq_length_5` and `seq_length_30`?

File: predict_pv_yield/data/dataloader.py [hunk 7]
```diff
@@ -141,26 +143,31 @@ def __getitem__(self, idx):
...
+            'gsp_id': torch.randn(self.batch_size, 32),
         }
 
         # add a nan
         x["pv_yield"][0, 0, :] = float("nan")
 
         # add fake x and y coords, and make sure they are sorted
-        x['sat_x_coords'], _ = torch.sort(torch.randn(self.batch_size, self.seq_length))
-        x['sat_y_coords'], _ = torch.sort(torch.randn(self.batch_size, self.seq_length), descending=True)
+        x['sat_x_coords'], _ = torch.sort(torch.randn(self.batch_size, self.seq_length_5))
```
Comment (JackKelly) [inline predict_pv_yield/data/dataloader.py:163]: Using `randn` is fine for now... But, in the future, it might be good to have `FakeDataset` spit out data which has the same statistical properties as the real data, because some downstream code might expect that.  For example, maybe the `sat_x_coords` and `nwp_x_coords` should be real OSGB coordinates, and should cover the same region of space.  But, yeah, maybe we shouldn't worry about this now!

File: predict_pv_yield/models/conv3d/model.py [hunk 8]
```diff
@@ -43,27 +43,29 @@ def __init__(
...
         """
 
-        self.forecast_len = forecast_len
-        self.history_len = history_len
         self.include_pv_yield = include_pv_yield
         self.include_nwp = include_nwp
         self.include_time = include_time
         self.number_of_conv3d_layers = number_of_conv3d_layers
-        self.number_of_nwp_features = 10*19*2*2
+        self.number_of_nwp_features = 10 * 19 * 2 * 2
```
Comment (JackKelly) [inline predict_pv_yield/models/conv3d/model.py:62]: Should this be configurable by the user?

**Review by JackKelly — APPROVED** [review]
Really exciting to see these experiments!  And I love your README notes!

File: predict_pv_yield/data/dataloader.py (hunk 6 above)
Comment (peterdudfield) [reply predict_pv_yield/data/dataloader.py:126]: #84 will sort this - ive started work [here](https://github.com/openclimatefix/predict_pv_yield/tree/issue/84-fake-dataset) on this

File: predict_pv_yield/data/dataloader.py (hunk 7 above)
Comment (peterdudfield) [reply predict_pv_yield/data/dataloader.py:163]: #84

File: predict_pv_yield/models/conv3d/model.py (hunk 8 above)
Comment (peterdudfield) [reply predict_pv_yield/models/conv3d/model.py:62]: #84
